Metrics customization guide with .NET 9 Advice API details (#5911)
open-telemetry/opentelemetry-dotnet: docs/metrics/customizing-the-sdk/README.md

        new ExplicitBucketHistogramConfiguration { Boundaries = Array.Empty<double>() })
```

#### Determining Explicit Buckets for Histograms

In OpenTelemetry .NET, histograms can be customized using both

> the View API and the Advice API. The actual explicit buckets used by

Reviewer: ```suggestion
the View API and the Advice API. The actual explicit buckets used by
```

the SDK are determined based on the following factors:

1. The explicit buckets provided by the instrumentation owner via the Advice API.
2. The explicit buckets defined through the View API during SDK initialization.
3. The default buckets provided by the SDK.

> Here’s how the SDK determines which explicit buckets to use:

Reviewer: ```suggestion
Here's how the SDK determines which explicit buckets to use:
```

1. **View API**: If custom buckets are defined via the View API, these
take precedence over any buckets provided by the Advice API or SDK defaults.
2. **Advice API**: If the View API does not specify custom buckets, the SDK
uses the explicit buckets provided by the Advice API.
3. **SDK Defaults**: If neither the View API nor the Advice API specifies
custom buckets, the SDK applies its default configuration.

Additionally, if an exponential histogram is defined using the View API,
the explicit bucket boundaries provided by the Advice API will be ignored,

> as exponential histograms take precedence.

Reviewer: Want to make sure I understand this - if the user leveraged the View API to choose a subset of the attributes, and there are no explicit buckets provided - would the buckets provided by the Hint API take effect or not?

Author: Yes, the buckets provided by the Advice (Hint) API would take effect in this case. Here's what I understand:

1. Exponential Histogram with View API: If the user defines an exponential histogram via the View API, it takes full precedence, and any explicit bucket boundaries provided by the Advice API are ignored.

2. No Explicit Buckets Provided by View API: If the View API is used but does not define explicit bucket boundaries (e.g., it only filters attributes or configures other aspects), then:

- The SDK will fall back to the explicit buckets from the Advice (Hint) API if they are available.

- If no buckets are defined in the Advice API either, then the SDK defaults apply.

@cijothomas Could you please verify if these are correct?

##### Example: Using Explicit Buckets via the Advice API

Here is an example of how to use the Advice API to specify explicit bucket

>  boundaries for a histogram:

Reviewer: ```suggestion
boundaries for a histogram:
```

        new ExplicitBucketHistogramConfiguration { Boundaries = Array.Empty<double>() })
```

### Customizing OpenTelemetry .NET SDK for Metrics

#### Using Explicit Buckets in Histograms

In OpenTelemetry .NET, histograms can be customized using the View API, and the Advice API allows additional flexibility for defining explicit bucket boundaries in instrumentation.

> ##### View API vs. Advice API: Defining Histogram Buckets

Reviewer: I understand where this is coming from. However, I suggest that we take a different approach, instead of having "vs." in our mind, focus on "how these things work together".

This is what I suggest:

1. The actual explicit buckets will be determined by the SDK based on the following factors:
    * The explicit buckets that the instrumentation owner provided via the Advice API.
    * The explicit buckets provided via View API during SDK initialization.
    * The default buckets provided by the SDK.
2. Here is how the SDK will determine the buckets:
    * ...

See [Program.cs](./Program.cs) for a complete example.

> ### Explicit Bucket Histogram Aggregation with .NET 9 Advice API

Reviewer 2: It’s misleading to call it .net 9 advice api as this api can be used in older .net runtimes

Reviewer 2: This is unresolved.

> ### Explicit Bucket Histogram Aggregation with .NET 9 Advice API

Reviewer 2: thus document is about sdk customization, so histogram advice is somewhat misfit here. we can mention it in the views section about custom buckets, but anything more should be elsewhere.

Reviewer 2: this is unresolved.

> ### Explicit Bucket Histogram Aggregation with .NET 9 Advice API

Reviewer: I doubt if this doc the right place, this doc is focusing on how to customize the SDK. Advice API is part of the instrumentation.

With the introduction of the .NET 9 Advice API, developers can
customize histogram metrics more effectively. The Advice API
allows for explicit bucket boundaries when defining histograms,
leading to more precise and informative metrics.

#### Overview of the .NET 9 Advice API

The .NET 9 Advice API provides enhancements for metrics collection,
enabling developers to specify how histograms should behave.
This feature is particularly useful when dealing with a known
distribution of values, as it allows for tailored metrics
collection that aligns with application requirements.

#### Using Explicit Bucket Boundaries

To use explicit bucket boundaries in your histograms, follow these steps:

1. **Install Required Packages**: Ensure you have the latest version of the
OpenTelemetry .NET SDK that supports the Advice API.

> 2. **Define Histogram with Explicit Buckets**: When creating a histogram metric,

Reviewer 2: One important information here is to document how view api vs advice interact, in particular the priority when both exist. 
I think advice api gets priority.
also mention that if exponential histogram is specified via views, then that takes priority and buckets from hint are ignored

Reviewer 3: Priority when selecting explicit histogram bounds:

* View API (always wins if specified)
* Advice API
* SDK defaults (used when nothing else is specified)

specify the bucket boundaries explicitly using the new API.

**Example**:

```csharp
using OpenTelemetry.Metrics;

// Create a meter
var meter = new Meter("MyApplication");

// Define bucket boundaries
double[] bucketBoundaries = { 0.0, 10.0, 20.0, 30.0, 40.0, double.MaxValue };

// Create a histogram with explicit buckets
var histogram = meter.CreateHistogram<double>("request_duration", "ms", "Measures the duration of requests", new HistogramOptions { BucketBoundaries = bucketBoundaries });

// Record values
histogram.Record(15.0);
histogram.Record(25.0);
```

#### Best Practices

- **Select Meaningful Bucket Boundaries**: When defining bucket
 boundaries, consider the typical range of values that your application
 processes. This helps to ensure that the histogram provides relevant
 insights.

> - **Monitor and Adjust**: After implementing explicit bucket

Reviewer 2: Without some concrete steps to monitor, I think it’s best to omit this.

Reviewer 2: @Annosha A lot of comments are marked resolved, but I don't see how. Please leave a comment about how the comment is addressed before resolving them.

OpenTelemetry SDK comes with the following `ExemplarFilter`s (defined on
`ExemplarFilterType`):

> - (Default behavior) `AlwaysOff`: Makes no measurements eligible for becoming an

Reviewer 2: please remove all unrelated changes from this PR. We are glad to accept them as a separate PR, so as to keep each PR focused on only one aspect only.
